sys/net/nfc/ndef/ndef_message.c

```diff
@@ -0,0 +1,270 @@
+    }
+    else {
+        #if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+            // payload length shall be MSB first or Big Endian - need to reverse if LE
+            uint32_t payload_length_multi_byte = htonl(payload_length);
+            record.payload_length = ndef_write_to_buffer(message, (uint8_t*) &payload_length_multi_byte, LONG_RECORD_PAYLOAD_LENGTH_SIZE);
+        #else        
+            record.payload_length = ndef_write_to_buffer(message, (uint8_t *)&payload_length, LONG_RECORD_PAYLOAD_LENGTH_SIZE);
+        #endif
+    }
+
+    if (id) {
+        record.id_length = ndef_write_to_buffer(message, &id_length, 1);
+    }
+
+    record.type = ndef_write_to_buffer(message, (uint8_t *)type, type_length);
+
+    if (id) {
+        record.id = ndef_write_to_buffer(message, id, id_length);
+    }
+
+    if (payload != NULL) {
+        record.payload = ndef_write_to_buffer(message, payload, payload_length);
+    }
+    else {
+        /** 
+         *  this needs to be done so the payload points to the correct position 
+         *  for further writes
+         */
+        record.payload = &message->buffer.memory[message->buffer.cursor];
+
+        /* the payload has to be written by the calling function */
+        (void)payload;
+    }
+
+    message->records[message->record_count] = record;
+    message->record_count += 1;
+
+    return 0;
+}
+
+int ndef_remove_record(ndef_t *message)
+{
+    assert(message != NULL);
+
+    if (message->record_count == 0) {
+        LOG_ERROR("No records to remove");
+        return -1;
+    }
+
+    /* for only one record we simply set the count and cursor to 0 */
+    if (message->record_count == 1) {
+        message->record_count = 0;
+        message->buffer.cursor = 0;
+        return 0;
+    }
+
+    /**
+     * the new cursor position can be found by looking at the start pointer
+     * of the last record and setting the cursor to that location 
+     */
+    ndef_record_desc_t *last_record = &message->records[message->record_count - 1];
+
+    /** 
+     * the new cursor position is the difference between the start of the first record
+     * record and the start of the last record as the cursor counts the bytes
+     */
+    uint32_t new_cursor = last_record->start - message->records[0].start;
+    message->buffer.cursor = new_cursor;
+    message->record_count -= 1;
+
+    /* the last record must now have ME to 1 */
+    message->records[message->record_count - 1].header[0] |= RECORD_ME_MASK;
+
+    return 0;
+}
+
+void ndef_clear(ndef_t *message)
+{
+    assert(message != NULL);
+
+    message->record_count = 0;
+    message->buffer.cursor = 0;
+}
+
+size_t ndef_calculate_record_size(uint8_t type_length, uint8_t id_length, uint32_t payload_length)
+{
+    size_t size = NDEF_RECORD_HEADER_SIZE + NDEF_RECORD_TYPE_LENGTH_SIZE + type_length;
+
+    if (id_length > 0) {
+        size += NDEF_RECORD_ID_LENGTH_SIZE + id_length;
+    }
```

Comment:
```suggestion
    size += NDEF_RECORD_ID_LENGTH_SIZE + id_length;
```

`id_length` is `uint8_t`, hence cannot be below zero.